### PR TITLE
Purge pipelines from distant history.

### PR DESCRIPTION
**Pull request by Ms2ger**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11893)

### CODE REVIEW

**Comment by asajeffrey** [comment]
cc @ConnorGBrewster

**Comment by paulrouget** [comment]
Can someone summarise what that does?

Is history still conserved even if pipelines are purged?
Are non-accessible pipelines (pipeline1 navigate to pipeline2, goBack to pipeline1, navigate to pipeline3, pipeline2 is not accessible anymore) discarded (maybe it was already)?

**Comment by ConnorGBrewster** [comment]
@paulrouget I believe the idea is that we will keep `3`(number can change) entries in the history, any pipelines past that will get closed. Once those pipelines are traversed to again, they are reloaded. This keeps us from having many many pipelines open after navigating a lot.

**Comment by paulrouget** [comment]
> Once those pipelines are traversed to again, they are reloaded

Ok, so we keep the history even if the pipeline is purged. That's all I was worried about :)

**Comment by ConnorGBrewster** [comment]
@paulrouget yup! they are reloaded here https://github.com/servo/servo/pull/11893/files#diff-55c92a6a5ba7654ce45fe6fc6c63740fR1416

**Comment by asajeffrey** [comment]
I'm quite concerned about what happens when pipelines share a script thread. If page A contains a same-origin iframe, which starts out at B1, then navigates to B2, B3 and B4. This will trigger `Pipeline::exit` on B1, which shares a script thread with A.

**Comment by asajeffrey** [comment]
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[components/constellation/constellation.rs, line 222 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11893#-KLMW_QBSFV5USfe1g2U:-KLMW_QBSFV5USfe1g2V:1883773357) ([raw file](https://github.com/servo/servo/blob/2f4953a0c10ef744f5d6858a6c3fbe7031e4f817/components/constellation/constellation.rs#L222)):*
> ```Rust
>         id: PipelineId,
>     },
>     Dead {
> ```

The terminology of alive vs dead doesn't seem aligned with the spec. https://html.spec.whatwg.org/multipage/browsers.html#session-history has a note saying "when a Document is not active, it's possible for it to be discarded to free resources," defined in https://html.spec.whatwg.org/multipage/browsers.html#discard-a-document.

---

*[components/constellation/constellation.rs, line 224 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11893#-KLMX6CgB4m8374QYFu2:-KLMX6ChOvWn6cvJTzJR:401122021) ([raw file](https://github.com/servo/servo/blob/2f4953a0c10ef744f5d6858a6c3fbe7031e4f817/components/constellation/constellation.rs#L224)):*
> ```Rust
>     Dead {
>         url: Url,
>         is_private: bool,
> ```

Should the data saved here be `LoadData`?

Is the privacy associated with the history entry or with the frame? (That is, can we have a frame, some of whose entries are private and some are public?)

---

*[components/constellation/constellation.rs, line 244 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11893#-KLMYwbFBPK3mPOCQqKS:-KLMYwbFBPK3mPOCQqKT:-2119976425) ([raw file](https://github.com/servo/servo/blob/2f4953a0c10ef744f5d6858a6c3fbe7031e4f817/components/constellation/constellation.rs#L244)):*
> ```Rust
>     }
> 
>     fn load(&mut self,
> ```

Can we add a comment saying what the semantics of this function is?

---

*[components/constellation/constellation.rs, line 246 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11893#-KLMYJ6JHwiDzAJM3f8U:-KLMYJ6JHwiDzAJM3f8V:502661033) ([raw file](https://github.com/servo/servo/blob/2f4953a0c10ef744f5d6858a6c3fbe7031e4f817/components/constellation/constellation.rs#L246)):*
> ```Rust
>     fn load(&mut self,
>             pipeline_id: PipelineId,
>             pipelines: &HashMap<PipelineId, Pipeline>,
> ```

Passing the pipelines hash map around seems like a break in the abstraction barrier between `Frame` and `Constellation`.

---

*[components/constellation/constellation.rs, line 247 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11893#-KLMYrqqGdx0DFyeXQEY:-KLMYrqqGdx0DFyeXQEZ:1571646736) ([raw file](https://github.com/servo/servo/blob/2f4953a0c10ef744f5d6858a6c3fbe7031e4f817/components/constellation/constellation.rs#L247)):*
> ```Rust
>             pipeline_id: PipelineId,
>             pipelines: &HashMap<PipelineId, Pipeline>,
>             through_history: bool)
> ```

What is this parameter for?

---

*[components/constellation/constellation.rs, line 264 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11893#-KLMZ3xQfVh2EQyGyY5l:-KLMZ3xQfVh2EQyGyY5m:483537222) ([raw file](https://github.com/servo/servo/blob/2f4953a0c10ef744f5d6858a6c3fbe7031e4f817/components/constellation/constellation.rs#L264)):*
> ```Rust
>     }
> 
>     fn maybe_purge(&mut self, pipelines: &HashMap<PipelineId, Pipeline>) {
> ```

Shouldn't this functionality be in `Constellation` rather than `Frame`?

---

*[components/constellation/constellation.rs, line 273 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11893#-KLMZIyxoobx3e4S8Ets:-KLMZIyxoobx3e4S8Ett:1578695497) ([raw file](https://github.com/servo/servo/blob/2f4953a0c10ef744f5d6858a6c3fbe7031e4f817/components/constellation/constellation.rs#L273)):*
> ```Rust
>                             None => continue,
>                         };
>                         pipeline.exit();
> ```

Doesn't this exit a pipeline even if the script thread is being used by its parent?

---

*[components/constellation/constellation.rs, line 1423 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11893#-KLMb3-OkKaaNI__cz_e:-KLMb3-P4XHD52f50LTE:-1833210932) ([raw file](https://github.com/servo/servo/blob/2f4953a0c10ef744f5d6858a6c3fbe7031e4f817/components/constellation/constellation.rs#L1423)):*
> ```Rust
>                     self.handle_load_start_msg(&pipeline_id);
>                     self.push_pending_frame(pipeline_id, Some(prev), true);
>                     self.compositor_proxy.send(ToCompositorMsg::ChangePageUrl(pipeline_id, url));
> ```

Should we set the window URL, even if it's a child iframe that caused this navigation?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11893)*

**Comment by asajeffrey** [comment]
As a general comment, it strikes me that this is one way of handling document discarding, but not the only way. Rather than having a per-frame hard limit on history, we could have a per-constellation soft limit.

**Comment by asajeffrey** [comment]
@Ms2ger is this still active?

**Comment by Ms2ger** [comment]
Yes, I'm going to get back to your review comments, hopefully this week.

**Comment by asajeffrey** [comment]
@Ms2ger ping?

**Comment by Ms2ger** [comment]
Rebased.

File: components/constellation/constellation.rs [hunk 1]
```diff
@@ -636,28 +721,28 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
         }
     }
 
-    fn joint_session_future(&self, frame_id_root: FrameId) -> Vec<(Instant, FrameId, PipelineId)> {
+    fn joint_session_future(&self, frame_id_root: FrameId) -> Vec<(Instant, FrameId, HistoryEntry)> {
         let mut future = vec!();
         for frame in self.full_frame_tree_iter(frame_id_root) {
-            future.extend(frame.next.iter().map(|entry| (entry.instant, entry.frame_id, entry.pipeline_id)));
+            future.extend(frame.next.iter().map(|entry| (entry.instant, entry.frame_id, entry.entry.clone())));
```
Comment (ConnorGBrewster) [inline components/constellation/constellation.rs]: nit: maybe we should mess with the naming here to avoid `entry.entry`.

**Comment by ConnorGBrewster** [comment]
Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[components/constellation/constellation.rs, line 305 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11893#-KRLVuN_0sHTFuD6E2VT:-KRLVuN_0sHTFuD6E2VU:b-ujkkpc) ([raw file](https://github.com/servo/servo/blob/0e5023b1b5fb2c261d3ba1db2bb1d1012a93c6e5/components/constellation/constellation.rs#L305)):*
> ```Rust
>     }
> 
>     fn maybe_purge(&mut self, pipelines: &HashMap<PipelineId, Pipeline>) {
> ```

Should we be purging everything but the previous/next `x` entries at the joint session history(top-level browsing context) level or at the individual session history/frame(browsing context) level?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11893)*

**Comment by Ms2ger** [comment]
Added some docs for `load()`; the rest of the comments seem like questions the two of you are more qualified to answer than I am.

**Comment by Ms2ger** [comment]
I have no idea what this is doing anymore. @asajeffrey @ConnorGBrewster, wanna take over?

**Comment by asajeffrey** [comment]
This is quite tricky, especially with sharing script threads (and there are bugs around when script threads are shared, sigh). We may need to postpone this until we get the sharing script threads story straight. Either that or we need to work out how to get a script thread to cope with some of its globals being dead, and some being alive.

**Comment by metajack** [comment]
@asajeffrey is going to take over this work. He's planning to rework script thread sharing which will further impact this, and at this point it probably makes sense for him to just include purging in his plans.

**Comment by asajeffrey** [comment]
@ConnorGBrewster: document unloading will be in our future :smile:

**Comment by asajeffrey** [comment]
#14211 fixes #633, so now I'm turning my attention to this.

**Pull request closed without merging**